pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_pullsecret_reconcile.go

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2022 Contributors to the Open Cluster Management project
+ */
+
+package klusterletcontroller
+
+import (
+	"context"
+	"fmt"
+	"github.com/openshift/library-go/pkg/operator/events"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	operatorapiv1 "open-cluster-management.io/api/operator/v1"
+	"open-cluster-management.io/registration-operator/pkg/helpers"
+)
+
+// pullSecretReconcile sync pull imagepullsecret to necessary namespaces on both managed and management clusters.
+type pullSecretReconcile struct {
+	managedClusterClients *managedClusterClients
+	kubeClient            kubernetes.Interface
+	operatorNamespace     string
+	recorder              events.Recorder
+}
+
+func (r *pullSecretReconcile) reconcile(ctx context.Context, klusterlet *operatorapiv1.Klusterlet, config klusterletConfig) (*operatorapiv1.Klusterlet, reconcileState, error) {
+	err := r.ensureNamespace(ctx, r.kubeClient, klusterlet, config.AgentNamespace)
+	if err != nil {
+		return klusterlet, reconcileStop, err
+	}
+
+	// Sync pull secret to the agent namespace
+	err = r.syncPullSecret(ctx, r.kubeClient, r.kubeClient, klusterlet, config.AgentNamespace, r.recorder)
+	if err != nil {
+		return klusterlet, reconcileStop, err
+	}
+
+	// For now, whether in Default or Hosted mode, the addons will be deployed on the managed cluster.
+	// sync image pull secret from management cluster to managed cluster for addon namespace
+	// TODO(zhujian7): In the future, we may consider deploy addons on the management cluster in Hosted mode.
+	addonNamespace := fmt.Sprintf("%s-addon", config.KlusterletNamespace)
+	// Ensure the addon namespace on the managed cluster
+	err = r.ensureNamespace(ctx, r.managedClusterClients.kubeClient, klusterlet, addonNamespace)
+	if err != nil {
+		return klusterlet, reconcileStop, err
+	}
+	// Sync pull secret to the klusterlet addon namespace
+	// The reason we keep syncing secret instead of adding a label to trigger addonsecretcontroller to sync is:
+	// addonsecretcontroller only watch namespaces in the same cluster klusterlet is running on.
+	// And if addons are deployed in default mode on the managed cluster, but klusterlet is deployed in hosted on management cluster, then we still need to sync the secret here in klusterlet-controller using `managedClusterClients.kubeClient`.
+	err = r.syncPullSecret(ctx, r.kubeClient, r.managedClusterClients.kubeClient, klusterlet, addonNamespace, r.recorder)
+	if err != nil {
+		return klusterlet, reconcileStop, err
+	}
+
+	if config.InstallMode == operatorapiv1.InstallModeHosted {
```

Comment:
Is it better to put this into the `managedReconcile`?

Reply:
yes, this reconcile is confusing, updated.